### ISSUE TITLE
mix of long and short queries make JDBC-driver hang

### ISSUE THREAD

**Issue by monetdb-team** [MonetDB/MonetDB#6693]
Date: 2019-03-20 17:42:32 +0100
From: Wouter &lt;<wouter>&gt;
To:   clients devs &lt;<bugs-clients>&gt;
Version: 11.31.13 (Aug2018-SP2)
CC:   martin.van.dinther, @PedroTadim

Last updated: 2019-09-26 17:19:39 +0200

## Comment 26936
Date: 2019-03-20 17:42:32 +0100
From: Wouter &lt;<wouter>&gt;

User-Agent:       Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:65.0) Gecko/20100101 Firefox/65.0
 Build Identifier: 
 
 We've been using the JDBC driver for years, but only recently we noticed a very nasty concurrency issue (the bug probably has been there all these years). We managed to reproduce it on various servers by just repeating the same test (sequentially, no parallelism). Some machines hang after a few attempts, some hang only after 1000's of attempts. It is not clear when a machine fails quickly, and when not. 
 
 The attached script triggers the issue eventually. The script intermittently does short and long queries (insert with blob of 9000 characters).
 
 Tested with monetdb-jdbc-2.28.
 
 
 Reproducible: Sometimes
 
 ### Steps to Reproduce:
 1. run script
 2. wait until it stops
 ### Actual Results:  
 it stops without completing, the jdbcclient is always at the same point:
 
 main:
   [1] java.net.SocketInputStream.socketRead0 (native method)
   [2] java.net.SocketInputStream.socketRead (SocketInputStream.java:116)
   [3] java.net.SocketInputStream.read (SocketInputStream.java:170)
   [4] java.net.SocketInputStream.read (SocketInputStream.java:141)
   [5] java.io.BufferedInputStream.fill (BufferedInputStream.java:246)
   [6] java.io.BufferedInputStream.read1 (BufferedInputStream.java:286)
   [7] java.io.BufferedInputStream.read (BufferedInputStream.java:345)
   [8] nl.cwi.monetdb.mcl.net.MapiSocket$BlockInputStream._read (null)
   [9] nl.cwi.monetdb.mcl.net.MapiSocket$BlockInputStream.readBlock (null)
   [10] nl.cwi.monetdb.mcl.net.MapiSocket$BlockInputStream.read (null)
   [11] sun.nio.cs.StreamDecoder.readBytes (StreamDecoder.java:284)
   [12] sun.nio.cs.StreamDecoder.implRead (StreamDecoder.java:326)
   [13] sun.nio.cs.StreamDecoder.read (StreamDecoder.java:178)
   [14] java.io.InputStreamReader.read (InputStreamReader.java:184)
   [15] java.io.BufferedReader.fill (BufferedReader.java:161)
   [16] java.io.BufferedReader.readLine (BufferedReader.java:324)
   [17] java.io.BufferedReader.readLine (BufferedReader.java:389)
   [18] nl.cwi.monetdb.mcl.io.BufferedMCLReader.readLine (null)
   [19] nl.cwi.monetdb.mcl.io.BufferedMCLReader.waitForPrompt (null)
   [20] nl.cwi.monetdb.jdbc.MonetConnection$ResponseList.executeQuery (null)
   [21] nl.cwi.monetdb.jdbc.MonetConnection$ResponseList.processQuery (null)
   [22] nl.cwi.monetdb.jdbc.MonetStatement.internalExecute (null)
   [23] nl.cwi.monetdb.jdbc.MonetStatement.execute (null)
   [24] nl.cwi.monetdb.client.JdbcClient.executeQuery (null)
   [25] nl.cwi.monetdb.client.JdbcClient.processInteractive (null)
   [26] nl.cwi.monetdb.client.JdbcClient.main (null)
 
 ### Expected Results:  
 It should complete the 100.000 runs without hanging.
 
 The script intermittently does short and long queries (insert with blob of 9000 characters). This makes it probably related to the SendThread, which is started when a query is longer than 8190 bytes.
 
 Note: severity is set to critical, as "the software hangs".

## Comment 26937
Date: 2019-03-20 17:43:31 +0100
From: Wouter &lt;<wouter>&gt;

Created attachment 614
 script to let jdbcclient hang (eventually)

> Attached file: [jdbc_bug_script.txt](https://github.com/MonetDB/monetdb-issues-attachments/blob/master/issue_6693_jdbc_bug_script.txt_614) (text/plain, 745 bytes)
> Description:   script to let jdbcclient hang (eventually)

## Comment 26938
Date: 2019-03-20 17:46:22 +0100
From: Wouter &lt;<wouter>&gt;

The attached script has been distilled from a java program. In the java program, we do an actual "conn.commit();" instead of running the "COMMIT;" SQL query. In that java program it always hangs at the following location:
 
 19 daemon prio=5 os_prio=0 tid=0x00007f0c144d7000 nid=0x1e runnable [0x00007f0bd41c6000]
    java.lang.Thread.State: RUNNABLE
 	at java.net.SocketInputStream.socketRead0(Native Method)
 	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
 	at java.net.SocketInputStream.read(SocketInputStream.java:171)
 	at java.net.SocketInputStream.read(SocketInputStream.java:141)
 	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
 	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
 	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
 	- locked <0x000000064793f030> (a java.io.BufferedInputStream)
 	at nl.cwi.monetdb.mcl.net.MapiSocket$BlockInputStream._read(Unknown Source)
 	at nl.cwi.monetdb.mcl.net.MapiSocket$BlockInputStream.readBlock(Unknown Source)
 	at nl.cwi.monetdb.mcl.net.MapiSocket$BlockInputStream.read(Unknown Source)
 	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
 	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
 	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
 	- locked <0x000000064794f0e0> (a java.io.InputStreamReader)
 	at java.io.InputStreamReader.read(InputStreamReader.java:184)
 	at java.io.BufferedReader.fill(BufferedReader.java:161)
 	at java.io.BufferedReader.readLine(BufferedReader.java:324)
 	- locked <0x000000064794f0e0> (a java.io.InputStreamReader)
 	at java.io.BufferedReader.readLine(BufferedReader.java:389)
 	at nl.cwi.monetdb.mcl.io.BufferedMCLReader.readLine(Unknown Source)
 	at nl.cwi.monetdb.mcl.io.BufferedMCLReader.waitForPrompt(Unknown Source)
 	- locked <0x000000064794f0a8> (a nl.cwi.monetdb.mcl.io.BufferedMCLReader)
 	at nl.cwi.monetdb.jdbc.MonetConnection$ResponseList.executeQuery(Unknown Source)
 	- locked <0x000000064793ed98> (a nl.cwi.monetdb.mcl.net.MapiSocket)
 	at nl.cwi.monetdb.jdbc.MonetConnection$ResponseList.processQuery(Unknown Source)
 	at nl.cwi.monetdb.jdbc.MonetConnection.sendTransactionCommand(Unknown Source)
 	at nl.cwi.monetdb.jdbc.MonetConnection.commit(Unknown Source)

## Comment 26939
Date: 2019-03-20 17:48:28 +0100
From: Wouter &lt;<wouter>&gt;

My analysis of the issue: the client seems to be waiting for new data from the server, while the server thinks it's already done.
 My guess it that it has to do with the code in MonetConnection.java around line 2704, but I also suspect there is something wrong in BufferedMCLWriter.writeLine. It sets a state in the reader (.setLineType(null)), while the two threads (reader and writer) don't really synchronize.

## Comment 27167
Date: 2019-07-24 17:28:52 +0200
From: MonetDB Mercurial Repository &lt;<hg>&gt;

Changeset 368078840ddf, made by Martin van Dinther <martin.van.dinther@monetdbsolutions.com> in the monetdb-java repo, refers to this bug.
 
 For complete details, see https://dev.monetdb.org/hg/monetdb-java?cmd=changeset;node=368078840ddf
 
 Changeset description:
 
 	Adding jdbc test for bugs 6517 and 6693 (hanging JDBC-driver problem)

## Comment 27169
Date: 2019-07-25 16:08:39 +0200
From: MonetDB Mercurial Repository &lt;<hg>&gt;

Changeset 637899bda602, made by Martin van Dinther <martin.van.dinther@monetdbsolutions.com> in the monetdb-java repo, refers to this bug.
 
 For complete details, see https://dev.monetdb.org/hg/monetdb-java?cmd=changeset;node=637899bda602
 
 Changeset description:
 
 	Removed SendThread class and its usage from MonetConnection.
 	It appeared the usage of a SendThread object (per connection) resulted sometimes to a non-responding/hanging JDBC connection, see bugs 6571 and 6693.
 	Tests with large query scripts and large results returned show that the SendThread class is not needed in the JDBC driver to function properly.
 	Also submitter of bug #6693 verified that the removal of SendThread resolved the hanging JDBC-driver problem without any side effects.

## Comment 27327
Date: 2019-09-26 17:19:39 +0200
From: Martin van Dinther &lt;<martin.van.dinther>&gt;

This is fixed in the new: monetdb-jdbc-2.29.jre7.jar release.
 It is downloadable from https://www.monetdb.org/downloads/Java/
